docs/development/logging.md

```diff
@@ -0,0 +1,174 @@
+  log := logf.FromContext(ctx)
+  log.Info("Reconciling Shoot")
+  // ...
+  return reconcile.Result{}, nil
+}
+```
+results in
+```text
+2021-12-16T09:35:59.099+0100    INFO    controller.shoot    Reconciling Shoot        {"name": "sunflower", "namespace": "garden-greenhouse"}
+```
+
+The logger is injected by controller-runtime's `Controller` implementation and our `controllerutils.CreateWorker` alike (if a logger is passed using `controllerutils.WithLogger`). The logger returned by `logf.FromContext` is never `nil`. If the context doesn't carry a logger, it falls back to the global logger (`logf.Log`), which might discard logs if not configured, but is also never `nil`.
+
+The controller implementation (controller-runtime / `CreateWorker`) itself takes care of logging the error returned by reconcilers.
+Hence, don't log an error that you are returning.
+Generally, functions should not return an error, if they already logged it, because that means the error is already handled and not an error anymore.
+See [Dave Cheney's post](https://dave.cheney.net/2015/11/05/lets-talk-about-logging) for more on this.
+
+### Messages
+
+- Log messages should be static. Don't put variable content in there, i.e., no `fmt.Sprintf` or string concatenation (`+`). Use key-value pairs instead.
+- Log messages should be capitalized. Note: this contrasts with error messages, that should not be capitalized. However, both should not end with a punctuation mark.
+
+### Keys and Values
+
+- Use `WithValues` instead of repeatedly adding key-value pairs for multiple log statements. `WithValues` creates a new logger from the parent, that carries the given key-value pairs. E.g., use it when acting on one object in multiple steps and logging something for each step:
+
+  ```go
+  log := parentLog.WithValues("infrastructure", client.ObjectKeyFromObject(infrastrucutre))
+  // ...
+  log.Info("Creating Infrastructure")
+  // ...
+  log.Info("Waiting for Infrastructure to be reconciled")
+  // ...
+  ```
+
+  Note: `WithValues` bypasses controller-runtime's special zap encoder that nicely encodes `ObjectKey`/`NamespacedName` and `runtime.Object` values, see [kubernetes-sigs/controller-runtime#1290](https://github.com/kubernetes-sigs/controller-runtime/issues/1290).
+  Thus, the end result might look different depending on the value and its `Stringer` implementation.
+
+- Use [lowerCamelCase](https://en.wiktionary.org/wiki/lowerCamelCase) for keys. Don't put spaces in keys, as it will make log processing with simple tools like `jq` harder.
+- Keys should be constant, human-readable, consistent across the codebase and naturally match parts of the log message, see [logr guideline](https://github.com/go-logr/logr#how-do-i-choose-my-keys).
+- When logging object keys (name and namespace), use the object's type as the log key (e.g. `pod`) and a `client.ObjectKey`/`types.NamespacedName` value as value (will be encoded into `{"namespace": "foo", "name": "bar"}`).
```

Comment:
I think showing an object name such as `foo/bar` (or just `foo` if it's a cluster object) is actually nicer than `{"namespace": "foo", "name": "bar"}`. I understand the motivation to have `{"namespace": "foo", "name": "bar"}` for the main object we are reconciling (these values are being added in `CreateWorker`, where we only have the request, but not the kind). For all other objects, perhaps `foo/bar` should be preferred, or at least allowed? We already have the utility function `kutil.ObjectName` that would properly return the full name depending on whether the object has a namespace or not.

Reply:
Hmm, I'm not sure why you would prefer `foo/bar` over properly structured logs.

Is it because you think, it is more effort? Passing an object key is just as easy as using `kutil.ObjectName`. Either you already have the key at hand or you can simply derive from the object you already have at hand e.g.:
```go
log.Info("Creating Deployment", "deployment", client.ObjectKeyFromObject(deployment))
```
This will result in something like
```text
{"level":"info","ts":"2021-12-16T08:32:21.059+0100","msg":"Creating Deployment","deployment":{"name": "bar", "namespace": "foo"}}
```
This is easier to process in logging stacks or using other tools like `jq`.

Or is it because you prefer the flat structure of `foo/bar` over a the structured log?

Reply:
My considerations were that it 1) "looks nicer" (for humans, not machines) and 2) it's already used a lot in existing logging, so it should rather be allowed.
However, I agree with the arguments why `{"namespace": "foo", "name": "bar"}` might be better for a "structured log". If we would like to discourage the use of `kutil.ObjectName` however, perhaps it's a good idea to add an explicit comment about it - otherwise it's likely that certain usages might survive the refactoring and transition to `logr` and then provide the wrong example for future code. What do you think?

Reply:
Ok, good point. I updated the document to explicitly discourage the use of `kutil.ObjectName`.